include/votca/csg/xyzreader.h

```diff
@@ -0,0 +1,150 @@
+ * limitations under the License.
+ *
+ */
+
+#ifndef __VOTCA_CSG_XYZREADER_H
+#define	__VOTCA_CSG_XYZREADER_H
+
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <votca/csg/topologyreader.h>
+#include <votca/csg/trajectoryreader.h>
+#include <votca/tools/constants.h>
+namespace votca { namespace csg {
+
+/**
+    \brief class for reading xyz files
+
+    This class provides the TrajectoryReader + Topology reader interface
+    for xyz files
+
+*/
+class XYZReader :
+    public TrajectoryReader, public TopologyReader
+{
+    public:
+        XYZReader() {}
+        ~XYZReader() {}
+
+        /// open a topology file
+        bool ReadTopology(std::string file, Topology &top);
+
+        /// open a trejectory file
+        bool Open(const std::string &file);
+        /// read in the first frame
+        bool FirstFrame(Topology &top);
+        /// read in the next frame
+        bool NextFrame(Topology &top);
+
+
+        template <class T>
+        void ReadFile(T &container){
+           if(!ReadFrame<true,T>(container)){
+               throw std::runtime_error("Reading xyz file failed");
+           }
+        }
+
+        void Close();
+
+    private:
+
+        template<class T>
+        int getContainerSize(T& container){
+            return container.size();
+        }
+
+        int getContainerSize(Topology& container){
+            return container.BeadCount();
+        }
+        
+        template <bool topology, class T>
+        void AddAtom(T &container,std::string name,int id, const tools::vec& pos){
+            Eigen::Vector3d pos2=pos.toEigen()*tools::conv::ang2bohr;
+            container.push_back(id,name,pos2);
+        }
+
+        template <bool topology, class T>
+        void AddAtom(Topology &container,std::string name,int id, const tools::vec& pos){
+        Bead *b;
+        tools::vec posnm=pos*tools::conv::ang2nm;
+        if(topology)
+            b = container.CreateBead(1, name+boost::lexical_cast<string>(id),
+                    (container.GetOrCreateBeadType(name)), 0, 0, 0);
+        else
+            b = container.getBead(id);
+            b->setPos(posnm);
+        }
+
+
+        template <bool topology, class T>
+        bool ReadFrame(T &container);
+
+        std::ifstream _fl;
+
+        int _line;
+};
+
+template <bool topology, class T>
+inline bool XYZReader::ReadFrame(T &container)
+{
+    string line;
+    getline(_fl, line); ++_line;
+    if(!_fl.eof()) {
+        // read the number of atoms
+        Tokenizer tok1(line," \t");
+        std::vector<std::string> line1;
+        tok1.ToVector(line1);
+        if(line1.size()!=1){
+          throw std::runtime_error("First line of xyz file should contain number of atoms/beads, nothing else.");
+        }
+        int natoms = boost::lexical_cast<int>(line1[0]);
+        if(!topology && natoms !=getContainerSize(container))
+            throw std::runtime_error("number of beads in topology and trajectory differ");
+
```

Comment:
Same as above.

Reply:
??

Reply:
My comments appeared in a different order. The point is to not rely on indentation but explicitly put brackets around the content in if loops etc...

Reply:
ahh.